File: Libraries/python/chardev.py
from pathlib import Path
from struct import Struct
from dataclasses import dataclass

CHAR_DEV_PATH = Path("/dev/char/")
INPUT_SYSFS = Path("/sys/class/input/")


@dataclass
class Timeval:
    tv_sec: int
    tv_usec: int


@dataclass
class InputKey:
    timeval: Timeval
    inp_type: int
    code: int
    value: int

    @staticmethod
    def data_format() -> Struct:
        return Struct("ddHHi")

    @classmethod
    def from_binary(cls, data: bytes):
        parsed = cls.data_format().unpack(data)

        return cls(Timeval(parsed[0], parsed[1]), parsed[2], parsed[3], parsed[4])

    def to_binary(self) -> bytes:
        return self.data_format().pack(
            self.timeval.tv_sec,
            self.timeval.tv_usec,
            self.inp_type,
            self.code,
            self.value,
        )

    @classmethod
    def with_freq(cls, freq: int):
        return cls(Timeval(0, 0), 0x12, 0x02, freq)


class CharDev:
    def __init__(self, dev: str) -> None:
        self.path = CHAR_DEV_PATH.joinpath(dev)

        if not self.path.exists():
            raise ValueError("Device does not exist")

    @classmethod
    def input_device_by_name(cls, name: str):
        assert INPUT_SYSFS.is_dir()

        for dev in INPUT_SYSFS.iterdir():
            name_p = dev.joinpath("device/name")

            if not name_p.is_file():
                continue

            with open(name_p, "r") as f:
                temp = f.read().strip()

                if temp != name:
                    continue

            dev_p = dev.joinpath("dev")
            with open(dev_p, "r") as f:
                temp = f.read().strip()
                return cls(temp)

        raise ValueError("Device not found")

    def read_evt(self) -> InputKey:
        with open(self.path, "rb") as f:
            data =  f.read(InputKey.data_format().size)
            return InputKey.from_binary(data)

    def write_evt(self, data: InputKey):
        with open(self.path, "wb") as f:
            f.write(data.to_binary())

File: Libraries/rust/beagle_helper/src/chardev.rs
use std::{
    fs::{File, OpenOptions},
    io::{self, Read, Write},
    path::Path,
};

const CHAR_DEV_PATH: &str = "/dev/char/";
const INPUT_SYSFS: &str = "/sys/class/input/";

pub struct CharDev(File);

impl CharDev {
    fn input_from_dev(dev: &str, opts: &OpenOptions) -> io::Result<Self> {
        let devp = Path::new(CHAR_DEV_PATH).join(dev);

        // Not sure why but the device is not a File. So cannot use [Path::is_file()]
        if devp.exists() {
            opts.open(devp).map(Self)
        } else {
            Err(io::Error::other("Device does not exist"))
        }
    }

    pub fn open_input_with_name(name: &str, opts: &OpenOptions) -> io::Result<Self> {
        let p = Path::new(INPUT_SYSFS);

        assert!(p.is_dir());
        for inp in std::fs::read_dir(p).unwrap() {
            let mut temp = String::with_capacity(10);
            let dev_p = inp.unwrap().path();

            let name_p = dev_p.join("device").join("name");

            // Some devices do not have `name` sysfs entry
            if !name_p.exists() {
                continue;
            }

            let mut name_f = File::open(&name_p).unwrap();
            name_f.read_to_string(&mut temp).unwrap();

            if temp.trim() == name {
                let mut temp = String::with_capacity(10);
                let mut devf = File::open(dev_p.join("dev")).unwrap();

                devf.read_to_string(&mut temp).unwrap();

                return Self::input_from_dev(temp.trim(), opts);
            }
        }

        Err(io::Error::new(
            io::ErrorKind::NotFound,
            "Failed to find any input named {name}",
        ))
    }

    pub fn write_evt(&mut self, evt: InputEvent) -> io::Result<()> {
        let bin: [u8; std::mem::size_of::<InputEvent>()] = unsafe { std::mem::transmute(evt) };
        self.0.write_all(&bin)
    }
}

impl io::Read for CharDev {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        self.0.read(buf)
    }
}

#[repr(C)]
#[derive(Default, Clone, Copy)]
pub struct Timeval {
    pub tv_sec: i64,
    pub tv_usec: i64,
}

/// Defined in [Event Interface](https://docs.kernel.org/input/input.html#event-interface)
#[repr(C)]
#[derive(Default, Clone, Copy)]
pub struct InputEvent {
    pub time: Timeval,
    pub r#type: std::ffi::c_ushort,
    pub code: std::ffi::c_ushort,
    pub value: std::ffi::c_int,
}

impl InputEvent {
    pub const fn with_frequency(freq: std::ffi::c_int) -> Self {
        Self {
            r#type: INPUT_TYPE_EV_SND,
            code: INPUT_CODE_SND_TONE,
            value: freq,
            time: Timeval {
                tv_sec: 0,
                tv_usec: 0,
            },
        }
    }
}

const INPUT_TYPE_EV_SND: std::ffi::c_ushort = 0x12;
const INPUT_CODE_SND_TONE: std::ffi::c_ushort = 0x02;
